### PR TITLE
AuthN: Rebuild Authenticate so we only have to call it once in context handler

### PR DESCRIPTION
**Pull request by kalleep**
**What is this feature?**
One of the goals with the new authn.Service is to be able to authenticate a request without having to specify a client.
To achieve this I created a new interface that clients can implement "ContextAwareClient". These type of clients need to be able to test if a request can be authenticated using that client and needs to specify a priority.

When registering new clients we can check if this additional interface is implement and add it to a simple priority list.
Then every time we call service.Authenticate we can go through the list in a consistent order.

The queue is implemented using generics so we can reuse the same structure for post auth hooks and post auth login.

**Special notes for your reviewer**:
* Need to add auth header to context in another pr

### CODE REVIEW

File: pkg/services/authn/authnimpl/service.go [hunk 1]
```diff
@@ -135,61 +139,73 @@ type Service struct {
...
+			if identity != nil {
+				return identity, nil
+			}
+		}
 	}
 
-	if !c.Test(ctx, r) {
-		return nil, false, nil
+	if latestErr != nil {
+		return nil, latestErr
```
[the remark this answers is not in the document]
Comment (kalleep) [reply pkg/services/authn/authnimpl/service.go:164]: How would they play with errutil.Error?

[the remark this answers is not in the document]
Comment (kalleep) [reply pkg/services/authn/authnimpl/service.go:164]: I guess it would unwrap to the first one

[the remark this answers is not in the document]
Comment (papagian) [reply pkg/services/authn/authnimpl/service.go:164]: good questions, let me try it

[the remark this answers is not in the document]
Comment (papagian) [reply pkg/services/authn/authnimpl/service.go]: the following test succeeds for me:
```
func TestMultierror(t *testing.T) {
	errInternal := NewBase(StatusInternal,
		"just.testing.internal", WithPublicMessage("internal error"))

	errUnauthorised := NewBase(StatusUnauthorized,
		"just.testing.unauthorised", WithPublicMessage("unauthorised"))

	var result error

	result = multierror.Append(result, errInternal.Errorf("err1"))
	result = multierror.Append(result, errUnauthorised.Errorf("err2"))

	assert.ErrorIs(t, result, errInternal)
	assert.ErrorIs(t, result, errUnauthorised)
	assert.Equal(t, result.Error(), "2 errors occurred:\n\t* [just.testing.internal] err1\n\t* [just.testing.unauthorised] err2\n\n")
}
```

[the remark this answers is not in the document]
Comment (kalleep) [reply pkg/services/authn/authnimpl/service.go]: Yeah It works, but when using error.As it seems to unwrap to the first one. And that is ok :)

File: pkg/services/authn/authnimpl/priority_queue.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package authnimpl
+
+type priority interface {
+	Priority() uint
+}
+
+func newQueue[T priority]() *queue[T] {
```
Comment (Jguer) [inline pkg/services/authn/authnimpl/priority_queue.go:7]:
```suggestion
func newQueue[T any]() *queue[T] {
```

File: pkg/services/authn/authnimpl/priority_queue.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	return &queue[T]{items: []T{}}
+}
+
+type queue[T priority] struct {
+	items []T
+}
+
+func (q *queue[T]) insert(c T) {
```
Comment (Jguer) [inline pkg/services/authn/authnimpl/priority_queue.go:15]:
```suggestion
func (q *queue[T]) insert(c T, priority int) {
```

maybe we can use it with `any` and without the priority method (defining the priority when we register them) so we can use this for post auth hooks as well?

**Review by Jguer — APPROVED** [review]

**Pull request merged**